tests/MaxMind/Test/WebService/ClientTest.php

```diff
@@ -15,33 +22,233 @@
  */
 class ClientTest extends TestCase
 {
-    public function test200(): void
+    /** @var Process */
+    public static $process;
+
+    /** @var int */
+    public static $port;
+
+    // Sets up the response that the test server is going to return.
+    /**
+     * addResponseInQueue.
+     *
+     * @param string $responseJSON the body that is going to be added into the queue
+     * @param int    $n            the number of times that it is going to be the response
+     */
+    public static function addResponseInQueue(string $responseJSON, $n = 1): void
+    {
+        if (!$fh = fopen(fullResponseFilePath, 'wb')) {
+            throw new \RuntimeException('Could not open tmp response json file.');
+        }
+        for ($n; $n > 0; $n--) {
+            fwrite($fh, $responseJSON . \PHP_EOL);
+        }
+        fclose($fh);
+    }
+
+    // Makes sure the built-in server is up by querying
+    // `/test` endpoint of the TestServer.
+    /**
+     * @return bool
+     */
+    public static function isWebsiteUp()
+    {
+        $requestUrl = 'localhost:' . (string) (self::$port) . '/test';
+
+        $ch = curl_init();
+
+        curl_setopt($ch, \CURLOPT_URL, $requestUrl);
+        curl_setopt($ch, \CURLOPT_HEADER, false);
+        curl_setopt($ch, \CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($ch, \CURLOPT_TIMEOUT, 2);
+        curl_exec($ch);
+
+        $response_status = curl_getinfo($ch, \CURLINFO_RESPONSE_CODE);
+
+        curl_close($ch);
+
+        return $response_status === 200;
+    }
+
+    public static function setUpBeforeClass(): void
+    {
+        // Clean up the response json if there is one.
+        if (file_exists(fullResponseFilePath)) {
+            unlink(fullResponseFilePath);
+        }
+
+        // Router is the test server controller
+        $routerPath = __DIR__ . '/TestServer.php';
+
+        // Getting a port that is available for use.
+        if (strtoupper(substr(\PHP_OS, 0, 3)) === 'WIN') {
+            //Windows
+            $socket = socket_create(\AF_INET, \SOCK_STREAM, \SOL_TCP);
+            socket_bind($socket, '0.0.0.0', 0);
+            socket_listen($socket);
+            socket_getsockname($socket, $addr, self::$port);
+            socket_close($socket);
```

Comment:
Is there a reason why the above wouldn't work on non-Windows OSes too? It looks like it might.

Reply:
I believe that wasn't working because of not having `socket` extension enabled in the config file. The conditional I put was left from my try when I didn't have `sockets` enabled. I have taken it back. Now, it should be working. (hopefully)